### ISSUE TITLE
wasi-sdk/bin/wasm-ld regression in 1.46.0

### ISSUE THREAD

**Issue by TjeuKayim** [rust-lang/rust#76679]
I tried to cross compile a hello world with WASI SDK as explained at https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/spec/wasm32_wasi/index.html:

```bash
cargo init
wget https://github.com/WebAssembly/wasi-sdk/releases/download/wasi-sdk-11/wasi-sdk_11.0_amd64_ubuntu20.04.deb
sudo dpkg -i wasi-sdk_11.0_amd64_ubuntu20.04.deb
rm wasi-sdk_11.0_amd64_ubuntu20.04.deb
export CC_wasm32_wasi=/opt/wasi-sdk/bin/clang
export CARGO_TARGET_WASM32_WASI_LINKER=/opt/wasi-sdk/bin/clang
cargo build --target wasm32-wasi
```

I expected to see this happen: the build succeeds with Rust v1.45.2 , so next versions should be compatible.

Instead, this happened when updating to Rust v1.46.0: https://github.com/TjeuKayim/wasi-link-fail/runs/1109104907

```
error: linking with `/opt/wasi-sdk/bin/clang` failed: exit code: 1
##[error]  |
  = note: wasm-ld: error: unknown argument: --eh-frame-hdr
          clang-10: error: linker command failed with exit code 1 (use -v to see invocation)
```

And 1.48.0-nightly (2020-09-12) produces another error: https://github.com/TjeuKayim/wasi-link-fail/runs/1109104926

```
error: linking with `/opt/wasi-sdk/bin/clang` failed: exit code: 1
##[error]  |
  = note: wasm-ld: error: std-5fceb8daecdd3ed8.std.5ro3oaeh-cgu.0.rcgu.o: Bad relocation type: 
          clang-10: error: linker command failed with exit code 1 (use -v to see invocation)
```

A weird exception is my Manjaro Linux system with Rust 1.46.0, it only builds succesfully when I add `RUSTFLAGS='-C link-args=-Wl,-v'`. I have no clue why, and couldn't reproduce this on Ubuntu.

See this repository for a minimal reproduction https://github.com/TjeuKayim/wasi-link-fail.

**Comment by mati865** [comment]
`--eh-frame-hdr` was recently fixed on nightly: https://github.com/rust-lang/rust/pull/76307

**Comment by LeSeulArtichaut** [comment]
@TjeuKayim is your issue fixed on the latest nightly?

**Comment by TjeuKayim** [comment]
> is your issue fixed on the latest nightly?

No, `rust version 1.48.0-nightly (6af1bdda5 2020-09-15)` is still affected.
I ran the Github Action again https://github.com/TjeuKayim/wasi-link-fail/runs/1122804715, the same error again:

```
  = note: wasm-ld: error: std-5fceb8daecdd3ed8.std.5ro3oaeh-cgu.0.rcgu.o: Bad relocation type: 
          clang-10: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by TjeuKayim** [comment]
The bisection below seems to point to the introduction of the other `eh-frame-hr` regression, that's useless. To bisect the other regression that causes this issue, it needs to apply the #76307 patch on rustc before testing, but I don't now how to execute such bisection.

searched nightlies: from nightly-2020-06-30 to nightly-2020-07-30
regressed nightly: nightly-2020-07-03
searched commits: from https://github.com/rust-lang/rust/commit/f781babf87dea29c44f93842b7ac9eb809549d29 to https://github.com/rust-lang/rust/commit/3503f565e1fb7296983757d2716346f48a4a262b
regressed commit: https://github.com/rust-lang/rust/commit/8a6d4342be6a6acbade8e7ef65e73d27ee8c9144

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.5.2</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --start=2020-06-30 --end=2020-07-30 --target wasm32-wasi -- build --target wasm32-wasi
```
</details>
